### PR TITLE
Refactor Format of Report File

### PR DESCRIPTION
**Pull request by t-yuminghe**

### CODE REVIEW

File: src/plugins/azgenerator/templates/topcommon/CliReport.ts [hunk 1]
```diff
@@ -7,162 +7,296 @@ import { CodeModelAz } from "../../CodeModelAz"
...
+            mo.push("");
+        } while (model.SelectNextCommand());
     }
-    mo.push("");
-
-    return mo;     
+    return mo;
 }
+
+// function getCommandBody1(model: CodeModelAz) {
```
Comment (LianwMS) [inline src/plugins/azgenerator/templates/topcommon/CliReport.ts:190]: Remove comment code

File: src/plugins/azgenerator/templates/topcommon/CliReport.ts [hunk 2]
```diff
@@ -7,162 +7,296 @@ import { CodeModelAz } from "../../CodeModelAz"
 import { ParameterLocation, SchemaType } from "@azure-tools/codemodel";
 import { isNullOrUndefined } from "util";
 
-export function GenerateAzureCliReport(model: CodeModelAz) : string[] {
+export function GenerateAzureCliReport(model: CodeModelAz): string[] {
     var output: string[] = [];
```
Comment (LianwMS) [inline src/plugins/azgenerator/templates/topcommon/CliReport.ts:11]: 1. May need rebase master as I move the file to templates/CliReports.ts. 
2. On the other hand, Could you use TemplateBase as base class?

File: src/plugins/azgenerator/templates/CliReport.ts [hunk 3]
```diff
@@ -1,168 +1,300 @@
...
  *  Copyright (c) Microsoft Corporation. All rights reserved.
  *  Licensed under the MIT License. See License.txt in the project root for license information.
  *--------------------------------------------------------------------------------------------*/
 
-import { SchemaType } from "@azure-tools/codemodel";
-import { isNullOrUndefined } from "util";
 import { CodeModelAz } from "../CodeModelAz";
+import { ParameterLocation, SchemaType } from "@azure-tools/codemodel";
+import { EscapeString,  ToCamelCase, Capitalize,  ToMultiLine, ToJsonString } from "../../../utils/helper"
+import { isNullOrUndefined } from "util";
```
Comment (LianwMS) [inline src/plugins/azgenerator/templates/CliReport.ts:9]: Is it possible to support generating report in incremental mode?

Comment (qiaozha) [reply src/plugins/azgenerator/templates/CliReport.ts:9]: The hidden command and commandgroup will be filtered by the Select method. the report generator doesn't need to change.

**Review by qiaozha — APPROVED** [review]

File: src/plugins/models.ts [hunk 4]
```diff
@@ -51,6 +51,7 @@ export class PathConstants {
     public static readonly actionFileOldVersion: string = "_actions.py";
     public static readonly initFile: string = "__init__.py";
     public static readonly helpFile: string = "_help.py";
+    public static readonly repotrFile: string = "report.md";
```
Comment (LianwMS) [inline src/plugins/models.ts:54]: >repotr [](start = 27, length = 6)

report

File: src/plugins/azgenerator/AzCoreFullGenerator.ts [hunk 5]
```diff
@@ -53,11 +53,13 @@ export class AzCoreFullGenerator extends AzGeneratorBase {
...
                 }
                 files[path.join(model.azOutputFolder, "manual/__init__.py")] = GenerateNamespaceInit(model);
-                await this.generateFullSingleAndAddtoOutput(new CliTopAction(model, isDebugMode));
-                await this.generateFullSingleAndAddtoOutput(new CliTopCustom(model, isDebugMode));
-                await this.generateFullSingleAndAddtoOutput(new CliTopInit(model, isDebugMode));
-                files[path.join(model.azOutputFolder, "report.md")] = GenerateAzureCliReport(model);
-                await this.generateFullSingleAndAddtoOutput(new CliMainDocSourceJsonMap(model, isDebugMode));
+                files[path.join(model.azOutputFolder, "action.py")] = await new CliTopAction(model, isDebugMode).fullGeneration();
+                files[path.join(model.azOutputFolder, "custom.py")] = await new CliTopCustom(model, isDebugMode).fullGeneration();
+                files[path.join(model.azOutputFolder, "__init__.py")] = await new CliTopInit(model, isDebugMode).fullGeneration();
```
Comment (LianwMS) [inline src/plugins/azgenerator/AzCoreFullGenerator.ts:58]: Keep the existing code

File: src/plugins/azgenerator/AzCoreFullGenerator.ts [hunk 6]
```diff
@@ -53,11 +53,13 @@ export class AzCoreFullGenerator extends AzGeneratorBase {
...
+                files[path.join(model.azOutputFolder, "report.md")] = await new CliReport(model, isDebugMode).fullGeneration();
```
Comment (LianwMS) [inline src/plugins/azgenerator/AzCoreFullGenerator.ts:59]: >files[path.join(model.azOutputFolder, "report.md")] = await new CliReport(model, isDebugMode).fullGeneration(); [](start = 16, length = 111)

try await this.generateFullSingleAndAddtoOutput().

File: src/plugins/azgenerator/AzCoreIncrementalGenerator.ts [hunk 7]
```diff
@@ -63,6 +64,10 @@ export class AzCoreIncrementalGenerator extends AzGeneratorBase {
         // Add Import from generated folder (Help)
         await this.generateIncrementalSingleAndAddtoOutput(new CliTopHelp(this.model, this.isDebugMode));
 
+        // Add Import from generated folder (Report)
+        const AzureCliReportGenerator = new CliReport(this.model, this.isDebugMode);
+        this.files[path.join(this.model.azOutputFolder, "report.md")] = await AzureCliReportGenerator.incrementalGeneration("");
```
Comment (LianwMS) [inline src/plugins/azgenerator/AzCoreIncrementalGenerator.ts:69]: await this.generateIncrementalSingleAndAddtoOutput(new CliReport(this.model, this.isDebugMode));

File: src/plugins/azgenerator/AzExtensionFullGenerator.ts [hunk 8]
```diff
@@ -56,12 +56,11 @@ export class AzExtensionFullGenerator extends AzGeneratorBase {
...
-        await this.generateFullSingleAndAddtoOutput(new CliTopAction(this.model, this.isDebugMode));
-        await this.generateFullSingleAndAddtoOutput(new CliTopCustom(this.model, this.isDebugMode));
-        await this.generateFullSingleAndAddtoOutput(new CliTopInit(this.model, this.isDebugMode));
-        await this.generateFullSingleAndAddtoOutput(new CliTopMetadata(this.model, this.isDebugMode));        
-
-        this.files["report.md"] = GenerateAzureCliReport(this.model);
+        this.files[this.azDirectory + "action.py"] = await new CliTopAction(this.model, this.isDebugMode).fullGeneration();
+        this.files[this.azDirectory + "custom.py"] = await new CliTopCustom(this.model, this.isDebugMode).fullGeneration();
+        this.files[this.azDirectory + "__init__.py"] = await new CliTopInit(this.model, this.isDebugMode).fullGeneration();
+        this.files[this.azDirectory + "azext_metadata.json"] = await new CliTopMetadata(this.model, this.isDebugMode).fullGeneration();
```
Comment (LianwMS) [inline src/plugins/azgenerator/AzExtensionFullGenerator.ts:62]: Keep the existing code

File: src/plugins/azgenerator/templates/CliReport.ts [hunk 9]
```diff
@@ -3,166 +3,207 @@
...
-            if (model.SelectFirstCommand())
-            {
-                do
-                {
-                    mo = getCommandBody(model);
-                    cmds[model.Command_Name] = mo;
+            do {
+    
+                /*mo.push("## " + model.Command_Name);
+                mo.push("");*/
```
Comment (LianwMS) [inline src/plugins/azgenerator/templates/CliReport.ts:64]: Remove

File: src/plugins/azgenerator/AzCoreFullGenerator.ts [hunk 10]
```diff
@@ -53,10 +53,10 @@ export class AzCoreFullGenerator extends AzGeneratorBase {
                     files[path.join(model.azOutputFolder, "vendored_sdks/__init__.py")] = GenerateNamespaceInit(model);
                 }
                 files[path.join(model.azOutputFolder, "manual/__init__.py")] = GenerateNamespaceInit(model);
-                await this.generateFullSingleAndAddtoOutput(new CliTopAction(model, isDebugMode));
+                await this.generateFullSingleAndAddtoOutput(new CliTopAction(model, isDebugMode));
```
Comment (LianwMS) [inline src/plugins/azgenerator/AzCoreFullGenerator.ts:56]: >	               [](start = 98, length = 15)

Remove spaces. or try to auto format code

File: src/plugins/azgenerator/AzExtensionFullGenerator.ts [hunk 11]
```diff
@@ -59,9 +59,8 @@ export class AzExtensionFullGenerator extends AzGeneratorBase {
         await this.generateFullSingleAndAddtoOutput(new CliTopAction(this.model, this.isDebugMode));
         await this.generateFullSingleAndAddtoOutput(new CliTopCustom(this.model, this.isDebugMode));
         await this.generateFullSingleAndAddtoOutput(new CliTopInit(this.model, this.isDebugMode));
-        await this.generateFullSingleAndAddtoOutput(new CliTopMetadata(this.model, this.isDebugMode));        
-
-        this.files["report.md"] = GenerateAzureCliReport(this.model);
+        await this.generateFullSingleAndAddtoOutput(new CliTopMetadata(this.model, this.isDebugMode));
+        this.files["report.md"] = await new CliReport(this.model, this.isDebugMode).fullGeneration();
```
Comment (LianwMS) [inline src/plugins/azgenerator/AzExtensionFullGenerator.ts:63]: >this.files["report.md"] = await new CliReport(this.model, this.isDebugMode).fullGeneration(); [](start = 8, length = 93)

using await this.generateFullSingleAndAddtoOutput()

File: src/plugins/azgenerator/AzExtensionIncrementalGenerator.ts [hunk 12]
```diff
@@ -62,6 +63,10 @@ export class AzExtensionIncrementalGenerator extends AzGeneratorBase {
         // Add Import from generated folder (Help)
         await this.generateIncrementalSingleAndAddtoOutput(new CliTopHelp(this.model, this.isDebugMode));
 
+        // Add Import from generated folder (Report)
+        const AzureCliReportGenerator = new CliReport(this.model, this.isDebugMode);
+        this.files["report.md"] = await AzureCliReportGenerator.incrementalGeneration("");
+
```
Comment (LianwMS) [inline src/plugins/azgenerator/AzExtensionIncrementalGenerator.ts:69]: using await this.generateIncrementalSingleAndAddtoOutput()

**Review by LianwMS — APPROVED** [review]
:shipit:

**Pull request merged**